src/pkg/runner/actions.go

```diff
@@ -0,0 +1,368 @@
+			// reference has already been processed.
+			if strings.Contains(task.Name, taskReferenceName+":") || strings.Contains(task.Name, actionReferenceName+":") {
+				return false
+			}
+		}
+		return true
+	}
+	return false
+}
+
+func getUniqueTaskActions(actions []types.Action) []types.Action {
+	uniqueMap := make(map[string]bool)
+	var uniqueArray []types.Action
+
+	for _, action := range actions {
+		if !uniqueMap[action.TaskReference] {
+			uniqueMap[action.TaskReference] = true
+			uniqueArray = append(uniqueArray, action)
+		}
+	}
+	return uniqueArray
+}
+
+func (r *Runner) performZarfAction(action *zarfTypes.ZarfComponentAction) error {
+	var (
+		ctx        context.Context
+		cancel     context.CancelFunc
+		cmdEscaped string
+		out        string
+		err        error
+
+		cmd = action.Cmd
+	)
+
+	// If the action is a wait, convert it to a command.
+	if action.Wait != nil {
+		// If the wait has no timeout, set a default of 5 minutes.
+		if action.MaxTotalSeconds == nil {
+			fiveMin := 300
+			action.MaxTotalSeconds = &fiveMin
+		}
+
+		// Convert the wait to a command.
+		if cmd, err = convertWaitToCmd(*action.Wait, action.MaxTotalSeconds); err != nil {
+			return err
+		}
+
+		// Mute the output because it will be noisy.
+		t := true
+		action.Mute = &t
+
+		// Set the max retries to 0.
+		z := 0
+		action.MaxRetries = &z
+
+		// Not used for wait actions.
+		d := ""
+		action.Dir = &d
+		action.Env = []string{}
+		action.SetVariables = []zarfTypes.ZarfComponentActionSetVariable{}
+	}
+
+	// load the contents of the env file into the Action + the RUN_ARCH
+	if r.envFilePath != "" {
+		envFilePath := filepath.Join(filepath.Dir(config.TaskFileLocation), r.envFilePath)
+		envFileContents, err := os.ReadFile(envFilePath)
+		if err != nil {
+			return err
+		}
+		action.Env = append(action.Env, strings.Split(string(envFileContents), "\n")...)
+	}
+
+	// load an env var for the architecture
+	action.Env = append(action.Env, fmt.Sprintf("%s_ARCH=%s", strings.ToUpper(config.EnvPrefix), config.GetArch()))
+
+	if action.Description != "" {
+		cmdEscaped = action.Description
+	} else {
+		cmdEscaped = message.Truncate(cmd, 60, false)
+	}
+
+	spinner := message.NewProgressSpinner("Running \"%s\"", cmdEscaped)
+	// Persist the spinner output so it doesn't get overwritten by the command output.
+	spinner.EnablePreserveWrites()
+
+	cfg := actionGetCfg(zarfTypes.ZarfComponentActionDefaults{}, *action, r.TemplateMap)
+
+	if cmd, err = actionCmdMutation(cmd); err != nil {
+		spinner.Errorf(err, "Error mutating command: %s", cmdEscaped)
+	}
+
+	// Template dir string
+	cfg.Dir = utils.TemplateString(r.TemplateMap, cfg.Dir)
+
+	// template cmd string
+	cmd = utils.TemplateString(r.TemplateMap, cmd)
+
+	duration := time.Duration(cfg.MaxTotalSeconds) * time.Second
+	timeout := time.After(duration)
+
+	// Keep trying until the max retries is reached.
+	for remaining := cfg.MaxRetries + 1; remaining > 0; remaining-- {
```

Comment:
It looks like there's a small oversight related to how the `context.Context` and its cancellation mechanism are managed, especially within the retry logic of our code. I noticed that the scope and handling of the `ctx` and `cancel` variables might not be aligned with the typical use-case scenarios, particularly within the loop structure.

The use of `defer cancel()` within the loop implies that the cancellation of contexts created for each retry attempt will be postponed until the entire `performZarfAction` function concludes. This approach might inadvertently lead to a couple of unintended consequences:

1. **Potential for Context Leaks:** The delay in calling `cancel()` after each loop iteration means that the resources associated with each context, like goroutines waiting for a context to be cancelled, aren't released immediately. This could result in unnecessary resource consumption until the function completes.
   
2. **Timeouts May Not Work As Expected:** Given that each command execution within the loop is intended to have a specific timeout, postponing the cancellation until the end of the function might cause individual commands to run longer than anticipated. This can affect the timing logic and potentially cause the function to run longer than it should, especially in scenarios where precise timeout control is crucial.

To address this, I suggest a minor adjustment to ensure that the `cancel()` function is invoked right after each attempt to execute a command, effectively within the loop. This means removing the `defer` statement in favor of directly calling `cancel()` after each command execution attempt, like so:

```go
for remaining := cfg.MaxRetries + 1; remaining > 0; remaining-- {
    ...
    select {
    case <-timeout:
        break
    default:
        ctx, cancel = context.WithTimeout(context.Background(), duration)
        if err := tryCmd(ctx); err != nil {
            cancel() // Directly cancel the context after an unsuccessful command attempt.
            continue
        }
        cancel() // Also cancel the context after a successful command attempt.
        return nil
    }
}
```

This adjustment ensures each context is properly and timely cancelled, aligning with the intended use of the context and cancel mechanisms, and maintaining resource efficiency and precise control over command execution timeouts.

A nice writeup about `defer gotchas` https://blog.learngoprogramming.com/gotchas-of-defer-in-go-1-8d070894cb01

Reply:
Thanks @naveensrinivasan ! I appreciate the detailed comments.